### ISSUE TITLE
Invariance not enforced for class-scoped TypeVar used in base class expression

### ISSUE THREAD

**Issue by erictraut** [python/mypy#13713]
Mypy does not detect the case where a covariant or contravariant class (i.e. a class that is parameterized by a class-scoped covariant or contravariant TypeVar) derives from an invariant class. This creates a type correctness hole as shown in the following code sample.

```python
from typing import TypeVar

T = TypeVar("T", covariant=True)

class MyList(list[T]): pass

def append_float(y: MyList[float]):
    y.append(1.0)

x: MyList[int] = MyList([1, 2, 3])
append_float(x)
```

Pyright likewise did not previously catch this case. I just added the check [here](https://github.com/microsoft/pyright/issues/3981).

**Comment by hauntsaninja** [comment]
It's gotten late for me, so I'm now confused by variance. I noticed you didn't implement the check for whether a covariant class derives from a contravariant class, and vice versa? I think we should do that as well, because it breaks transitivity of subtyping.

```
from typing import TypeVar, Generic
T = TypeVar('T')
T_co = TypeVar('T_co', covariant=True)
T_contra = TypeVar('T_contra', contravariant=True)

class A: ...
class B(A): ...

class Parent(Generic[T_co]): ...
class Child(Parent[T_contra], Generic[T_contra]): ...

def takes_parent_a(f: Parent[A]): ...
def takes_parent_b(f: Parent[B]): ...
def takes_child_a(f: Child[A]): ...
def takes_child_b(f: Child[B]): ...

def give_child_a(t: Child[A]):
    takes_parent_a(t)
    takes_parent_b(t)
    takes_child_a(t)
    takes_child_b(t)

def give_child_b(t: Child[B]):
    takes_parent_a(t)
    takes_parent_b(t)
    takes_child_a(t)
    takes_child_b(t)
```
I think both mypy and pyright agree that `Child[A]` is a `Child[B]` and `Child[B]` is a `Parent[B]`, but that `Child[A]` is not a `Parent[B]`.

It looks like typeshed breaks this with Container and SupportsGetItem...
```
from _typeshed import SupportsGetItem
from typing import Container

class A: ...
class B(A): ...

sgia: SupportsGetItem[A, None]
sgib: SupportsGetItem[B, None] = sgia
conb: Container[B] = sgib
conb = sgia  # boom
```
cc @JelleZijlstra

**Comment by ilevkivskyi** [comment]
Isn't this a duplicate of a (super-old) https://github.com/python/mypy/issues/736?

**Comment by erictraut** [comment]
@hauntsaninja, yes, I agree that mismatched covariance/contravariance mismatch is also a problem. I think it's unlikely that anyone would attempt to use those combinations, but for completeness that check could be added.

@ilevkivskyi, yes, this looks like a duplicate. Apologies, I didn't notice the existing bug. I'll close this one accordingly.
